engine/access/rpc/backend/backend_stream_transactions.go

```diff
@@ -5,307 +5,231 @@ import (
+	if err != nil {
+		b.log.Err(err).Str("block_id", startBlockID.String()).Msg("failed to get start height")
+		return subscription.NewFailedSubscription(err, "failed to get start height")
 	}
 
-	txInfo := transactionSubscriptionMetadata{
-		TransactionResult: &access.TransactionResult{
-			TransactionID: txID,
-			BlockID:       flow.ZeroID,
-			Status:        initialStatus,
-		},
-		txReferenceBlockID:   referenceBlockID,
-		blockWithTx:          nil,
-		eventEncodingVersion: requiredEventEncodingVersion,
-		shouldTriggerPending: shouldTriggerPending,
+	// Retrieve the current state of the transaction.
+	txInfo, err := newTransactionSubscriptionMetadata(ctx, b.backendTransactions, txID, referenceBlockID, requiredEventEncodingVersion)
+	if err != nil {
+		b.log.Err(err).Str("tx_id", txID.String()).Msg("failed to get current transaction state")
+		return subscription.NewFailedSubscription(err, "failed to get tx reference block ID")
 	}
 
-	return b.subscriptionHandler.Subscribe(ctx, nextHeight, b.getTransactionStatusResponse(&txInfo))
+	return b.subscriptionHandler.Subscribe(ctx, startHeight, b.getTransactionStatusResponse(txInfo))
 }
 
 // getTransactionStatusResponse returns a callback function that produces transaction status
 // subscription responses based on new blocks.
-func (b *backendSubscribeTransactions) getTransactionStatusResponse(txInfo *transactionSubscriptionMetadata) func(context.Context, uint64) (interface{}, error) {
+func (b *backendSubscribeTransactions) getTransactionStatusResponse(
+	txInfo *transactionSubscriptionMetadata,
+) func(context.Context, uint64) (interface{}, error) {
+	triggerMissingStatusesOnce := atomic.NewBool(false)
+
 	return func(ctx context.Context, height uint64) (interface{}, error) {
 		err := b.checkBlockReady(height)
 		if err != nil {
 			return nil, err
 		}
 
-		if txInfo.shouldTriggerPending {
-			return b.handlePendingStatus(txInfo)
-		}
-
-		if b.isTransactionFinalStatus(txInfo) {
-			return nil, fmt.Errorf("transaction final status %s already reported: %w", txInfo.Status.String(), subscription.ErrEndOfData)
+		if triggerMissingStatusesOnce.CompareAndSwap(false, true) {
+			return b.generateResultsStatuses(txInfo.txResult, flow.TransactionStatusUnknown)
 		}
 
-		// If on this step transaction block not available, search for it.
-		if txInfo.blockWithTx == nil {
-			// Search for transaction`s block information.
-			txInfo.blockWithTx,
-				txInfo.BlockID,
-				txInfo.BlockHeight,
-				txInfo.CollectionID,
-				err = b.searchForTransactionBlockInfo(height, txInfo)
-
-			if err != nil {
-				if errors.Is(err, storage.ErrNotFound) {
-					return nil, fmt.Errorf("could not find block %d in storage: %w", height, subscription.ErrBlockNotReady)
-				}
-
-				if !errors.Is(err, ErrTransactionNotInBlock) {
-					return nil, status.Errorf(codes.Internal, "could not get block %d: %v", height, err)
-				}
-			}
+		if txInfo.txResult.IsFinal() {
```

Comment:
we should also handle the case where the tx requested is not found within N blocks of subscribing. I think there are 2 ways we could handle it:
1. use the current sealed height when the stream is created as the placeholder for reference block. then the derived status lookup will automatically mark the tx expired after 600 blocks.
2. add a check here and return an error if tx status is unknown and it's been more than N blocks since the stream started. N could be something a little larger than the tx expiration to allow for some network jitter.

I think 2 would be the clearest for users since it more accurately communicates what happened.